ui/src/app/pages/analysis/runtime-panel.tsx

```diff
@@ -71,186 +73,201 @@ export interface Props {
+        const promise = runtimeApi().getRuntime(namespace, {signal: aborter.signal});
+        updateRuntimeOpsStoreForWorkspaceNamespace(namespace, {
           promise: promise,
           operation: 'get',
-          aborter: this.aborter
+          aborter: aborter
         });
-        runtime = await promise;
       } catch (e) {
         // 404 is expected if the runtime doesn't exist, represent this as a null
         // runtime rather than an error mode.
         if (e.status !== 404) {
-          error = true;
+          setError(true);
         }
+      } finally {
+        setLoading(false);
       }
-      markRuntimeOperationCompleteForWorkspace(this.props.workspace.namespace);
-      this.setState({
-        runtime,
-        error,
-        loading: false
-      });
-    }
-
-    render() {
-      const {runtimeOps, workspace} = this.props;
-      const {loading, error, runtime} = this.state;
+      markRuntimeOperationCompleteForWorkspace(namespace);
+    };
 
-      const activeRuntimeOp: RuntimeOperation = runtimeOps.opsByWorkspaceNamespace[workspace.namespace];
-
-      if (loading) {
-        return <Spinner style={{width: '100%', marginTop: '5rem'}}/>;
-      } else if (error) {
-        return <div>Error loading compute configuration</div>;
-      } else if (!runtime) {
-        // TODO(RW-5591): Create runtime page goes here.
-        return <React.Fragment>
-          <div>No runtime exists yet</div>
-          {activeRuntimeOp && <hr/>}
-          {activeRuntimeOp && <div>
-            <ActiveRuntimeOp operation={activeRuntimeOp.operation} workspaceNamespace={workspace.namespace}/>
-          </div>}
-        </React.Fragment>;
-      }
+    loadRuntime();
+    return () => aborter.abort();
+  });
 
-      const isDataproc = !!runtime.dataprocConfig;
+  if (loading) {
+    return <Spinner style={{width: '100%', marginTop: '5rem'}}/>;
+  } else if (error) {
+    return <div>Error loading compute configuration</div>;
+  } else if (!currentRuntime) {
+    // TODO(RW-5591): Create runtime page goes here.
+    return <React.Fragment>
+      <div>No runtime exists yet</div>
+      {activeRuntimeOp && <hr/>}
+      {activeRuntimeOp && <div>
+        <ActiveRuntimeOp operation={activeRuntimeOp.operation} workspaceNamespace={workspace.namespace}/>
+      </div>}
+    </React.Fragment>;
+  }
 
-      let masterMachineName;
-      let masterDiskSize;
-      if (isDataproc) {
-        masterMachineName = runtime.dataprocConfig.masterMachineType;
-        masterDiskSize = runtime.dataprocConfig.masterDiskSize;
-      } else {
-        masterMachineName = runtime.gceConfig.machineType;
-        masterDiskSize = runtime.gceConfig.bootDiskSize;
-      }
-      const machineType = allMachineTypes.find(({name}) => name === masterMachineName) || defaultMachineType;
+  const isDataproc = (currentRuntime && !!currentRuntime.dataprocConfig);
+  const runtimeChanged = selectedMachine || selectedDiskSize;
```

Comment:
Won't this still be true if the user sets it to a different machine or disk size and then sets it back to what it was before?

Reply:
The machine and disk selectors will only update with a value if the value has changed from the initial value, otherwise they will be set to null. I changed the name to `updatedMachine` and `updatedDiskSize` - does that make it any clearer in the code?